kratos/spatial_containers/geometrical_objects_bins.h

```diff
@@ -0,0 +1,472 @@
+            for(std::size_t k = min_position[2] ; k < max_position[2] ; k++){
+                for(std::size_t j = min_position[1] ; j < max_position[1] ; j++){
+                    for(std::size_t i = min_position[0] ; i < max_position[0] ; i++){
+                        auto cell_bounding_box = GetCellBoundingBox(i,j,k);
+                        if(IsIntersected(i_geometrical_object->GetGeometry(), cell_bounding_box, Tolerance)){
+                            GetCell(i,j,k).push_back(&(*i_geometrical_object));
+                        }
+                    }
+                }
+            }
+        }        
+    }
+
+    /// Giving the min and max position of cells intersecting with the bounding box of the geometry.
+    template<typename TGeometryType>
+    void CalculateMinMaxPositions(TGeometryType const& TheGeometry, array_1d< std::size_t, 3 >& MinPosition, array_1d< std::size_t, 3 >& MaxPosition){
+        if(TheGeometry.empty())
+            return;
+
+        BoundingBox<Point> bounding_box(TheGeometry.begin(), TheGeometry.end());
+
+        for(int i = 0; i < 3; i++ ) {
+            MinPosition[ i ] = CalculatePosition( bounding_box.GetMinPoint()[i], i );
+            MaxPosition[ i ] = CalculatePosition( bounding_box.GetMaxPoint()[i], i ) + 1;
+        }
+    }
+
+    /// calculating the cell position of a given coordinate in respective axis given by dimension
+    std::size_t CalculatePosition( double Coordinate, int ThisDimension ) const {
+        auto distance = Coordinate - mBoundingBox.GetMinPoint()[ ThisDimension ];
+        distance = ( distance < 0.00 ) ? 0.00 : distance;
+        std::size_t position =
+            static_cast< std::size_t >( distance * mInverseOfCellSize[ ThisDimension ] );
+        std::size_t result= ( position > mNumberOfCells[ ThisDimension ] - 1 )
+                                ? mNumberOfCells[ ThisDimension ] - 1
+                                : position;
+        return result;
+    }
+    
+    /// Expands by the tolerance the geometry bounding box and checks the intersection
+    template<typename TGeometryType>
+    static inline bool IsIntersected(TGeometryType& TheGeometry, BoundingBox<Point> Box, const double tolerance)
+    {
+        Point rLowPointTolerance;
+        Point rHighPointTolerance;
+        
+        for(std::size_t i = 0; i<3; i++)
+        {
+            rLowPointTolerance[i]  =  Box.GetMinPoint()[i] - tolerance;
+            rHighPointTolerance[i] =  Box.GetMaxPoint()[i] + tolerance;
+        }
+        
+        return  TheGeometry.HasIntersection(rLowPointTolerance,rHighPointTolerance);
+    }
+
+    /// Searchs in objects in the given cell for the ones with distance less than given radius + tolerance.
+    template<typename TPointType>
+    void SearchInRadiusInCell(CellType const& TheCell, TPointType const& ThePoint, double Radius, std::unordered_set<GeometricalObject*>& rResults) {
+        for(auto p_geometrical_object : TheCell){  
```

Comment:
Would it pay off to check if the cell is completely inside/outside the radius before looping the contained geometrical objects?

Reply:
Definitely! Let me see if I can deduce it from the in-out function of the cell.

Reply:
hmmm, then I cannot calculate the distance to the objects... Now I'm not sure if it worth to have it calculating and calculating the distance separately.

Reply:
I didn't notice you are calculating the distances of all neighbors - then i think it does not really pay off.